pkg/firewall/cache.go

```diff
@@ -0,0 +1,77 @@
+package firewall
+
+import (
+	"container/list"
+	"sync"
+	"time"
+)
+
+// timeCache provides a time cache safe for concurrent use by
+// multiple goroutines without additional locking or coordination.
+type timeCache struct {
+	indexer  *list.List
+	cache    map[string]time.Time
+	timespan time.Duration
+	mutex    sync.RWMutex
+}
+
+// NewTimeCache creates a new cache instance with provided timespan.
+func newTimeCache(timespan time.Duration) *timeCache {
+	return &timeCache{
+		indexer:  list.New(),
+		cache:    make(map[string]time.Time),
+		timespan: timespan,
+	}
+}
+
+// Add adds an entry to the cache. Returns `true` if entry was not present in
+// the cache and was successfully added into it. Returns `false` if
+// entry is already in the cache. This method is synchronized.
+func (tc *timeCache) add(item string) bool {
+	tc.mutex.Lock()
+	defer tc.mutex.Unlock()
+
+	_, ok := tc.cache[item]
+	if ok {
+		return false
+	}
+
+	// sweep old entries
+	for {
+		back := tc.indexer.Back()
+		if back == nil {
+			break
+		}
+
+		item := back.Value.(string)
+		itemTime, ok := tc.cache[item]
+		if !ok {
+			logger.Fatalf(
```

Comment:
It's a tough decision. It's not generally possible the cache can be inconsistent if there are no bugs in the implementation so this logger should be never printed.

I am fine with both. Fatal printed to the logs multiple times should wake up the operator anyway.

Reply:
> Fatal printed to the logs multiple times should wake up the operator anyway.

Just to confirm we are on the same page, fatal is killing the process, right?

Reply:
I think we are not on the same page :)

You asked
> do we want to kill the client in case of this error?

And my answer was *I do not know* while the current code here is NOT killing the process. We are logging on a `FATAL` level but we are not panicking. 

The severity is:
```
DEBUG < INFO < WARN < ERROR < DPANIC < PANIC < FATAL
```

When I look at it, I think we probably should change this log level to panic from fatal but I think we should avoid panicking here. It does not stop the application from processing, firewall checks may just get more expensive. I think it's better than killing the client and risking slashing.

Reply:
Updated in https://github.com/keep-network/keep-ecdsa/pull/406/commits/c5823ee900043c50a46058a11bbf02aab82c0a6a

Reply:
> And my answer was I do not know while the current code here is NOT killing the process. We are logging on a FATAL level but we are not panicking.

We are logging with `Fatalf` which as per [documentation](https://godoc.org/github.com/op/go-logging#Logger.Fatalf) is killing the client.
> Fatalf is equivalent to l.Critical followed by a call to os.Exit(1). 

---

> When I look at it, I think we probably should change this log level to panic from fatal but I think we should avoid panicking here.

So do you want to panic or not? 

---

> It does not stop the application from processing, firewall checks may just get more expensive. I think it's better than killing the client and risking slashing.

Panic is also killing the client.

---

The difference between `Panic` and `Fatal` will be in handling defers. But as I understand the client will end up closed in each case.

See: https://stackoverflow.com/a/35997323

Reply:
Good catch. I was looking at `ipfs/go-log` docs in the source code and it's not mentioned there so I assumed logger is just writing the message to the correct level and it's up to the programmer to decide what to do next. I experimented with it and `logger.Fatalf` results in:
```
Starting keep-ecdsa client...
14:58:15.310 CRITI   keep-cmd: inconsistent cache state - expected item [123] is not present start.go:90
```

While `logger.Panicf` results in:
```
14:54:16.769 CRITI   keep-cmd: inconsistent cache state - expected item [123] is not present start.go:90
panic: inconsistent cache state - expected item [123] is not present

goroutine 1 [running]:
github.com/whyrusleeping/go-logging.(*Logger).Panicf(0xc00011daa0, 0x4b5bc63, 0x3c, 0xc000221810, 0x1, 0x1)
        /Users/piotr/go/pkg/mod/github.com/whyrusleeping/go-logging@v0.0.0-20170515211332-0457bb6b88fc/logger.go:206 +0x145
github.com/keep-network/keep-ecdsa/cmd.Start(0xc00016a2c0, 0x0, 0xc00016a2c0)
        /Users/piotr/go/src/github.com/keep-network/keep-ecdsa/cmd/start.go:90 +0x1a0
github.com/urfave/cli.HandleAction(0x49b1080, 0x4cbfcd0, 0xc00016a2c0, 0xc0004aa200, 0x0)
        /Users/piotr/go/pkg/mod/github.com/keep-network/cli@v1.20.0/app.go:490 +0xc8
github.com/urfave/cli.Command.Run(0x4b26082, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x4b52aec, 0x2f, 0x0, ...)
        /Users/piotr/go/pkg/mod/github.com/keep-network/cli@v1.20.0/command.go:210 +0x9e8
github.com/urfave/cli.(*App).Run(0xc0001201a0, 0xc000138000, 0x4, 0x4, 0x0, 0x0)
        /Users/piotr/go/pkg/mod/github.com/keep-network/cli@v1.20.0/app.go:255 +0x741
main.main()
        /Users/pi:53 +0x3a2
```

I'd suggest we go with `Errorf` in this case. I am not happy with it because I think it's worse than ERROR but at the same time, I don't think we should `panic` here on production.

Reply:
So maybe [`Criticalf`](https://github.com/op/go-logging/blob/master/logger.go#L203)?

Reply:
I'd love to but there is no `Criticalf` in `ipfs` logger: https://github.com/ipfs/go-log/blob/master/log.go#L14